// kwill/webpack.config.js
'use strict';

const webpack = require('webpack'),
    path = require('path');


//  babel options

var babelOptions = {
    "presets": [
        ["es2015", {"modules": false}]
    ]
};

module.exports =
    {
        cache: true,
        entry: {
            main: './src/ts/index.ts'
        },
        output: {
            library: "Kwill",
            libraryTarget: "umd",
            umdNamedDefine: true,
            filename: './bin/kwill.js'
        },

        module: {
            rules: [
                {
                    test: /\.ts(x?)$/,
                    exclude: /node_modules/,
                    use: [
                        {
                            loader: 'babel-loader',
                            options: babelOptions
                        },
                        {
                            loader: 'ts-loader'
                        }
                    ]
                }
            ]
        },
        resolve: {
            extensions: ['.ts', '.tsx', '.js']
        }
    };
