// kwill/lib/text-format-ui/ui-components/ColourPicker.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var log = require("loglevel");
var Constants_1 = require("../../constants/Constants");
var FormatText_1 = require("../../editor/text-format/FormatText");
var kulr_1 = require("kulr");
function setupColourPicker(frame, options) {
    var colourPickerOptions = {
        id: options.id + '-colour-picker',
        elements: {
            selector: frame.contentDocument.querySelector('.colour-picker')
        },
        actions: {
            liveUpdate: false
        },
        defaultColour: 'aaaaaa'
    };
    try {
        var colourPicker = kulr_1.createColourPicker(colourPickerOptions);
        if (!colourPicker) return;
        colourPicker.observableColour.subscribe(function (hex) {
            return FormatText_1.formatText(Constants_1.KWILL_CONSTS.textFormats.TEXT_COLOUR, hex);
        });
        return colourPicker;
    } catch (e) {
        log.warn("can't create colour picker");
    }
}
exports.setupColourPicker = setupColourPicker;
