// kwil/dist/funder/escrowV6.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EscrowV6 = void 0;
const ethers_1 = require("ethers");
const override_1 = require("./override");
class EscrowV6 {
    constructor(validatorAddress, poolAddress, abi, signer) {
        this.contract = new ethers_1.ethers.Contract(poolAddress, abi, signer);
        this.signer = signer;
        this.validatorAddress = validatorAddress;
    }
    getTokenAddress() {
        if (!this.tokenAddress) {
            this.tokenAddress = this.contract['escrowToken()']();
        }
        return this.tokenAddress;
    }
    getDepositedBalance(address) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.contract['pools(address, address)'](this.validatorAddress, address);
        });
    }
    createOverride(method, args) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, override_1.createOverride)(this.signer, this.contract, method, args);
        });
    }
    deposit(amount, override) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!override) {
                override = this.createOverride('deposit', [this.validatorAddress, amount]);
            }
            return this.contract["deposit(address validator, uint256 amt)"](this.validatorAddress, amount, override);
        });
    }
}
exports.EscrowV6 = EscrowV6;
