// kwill/lib/text-format-ui/FrameUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var lodash_1 = require("lodash");
var FontLoader_1 = require("../utils/dom/style/FontLoader");
function isFramed(target) {
    return !target || target.tagName === 'IFRAME';
}
exports.isFramed = isFramed;
function captureElement(target, options, frameReadyCallback) {
    var iframe = document.createElement('iframe');
    iframe.style.setProperty('position', 'absolute');
    iframe.style.setProperty('left', '-9999px');
    iframe.setAttribute('class', target.getAttribute('class'));
    iframe.onload = function () {
        console.log('on frame load');
        populateFrame(iframe, target, options);
        frameReadyCallback(iframe);
    };
    target.parentElement.insertBefore(iframe, target);
    console.log('returning frame');
    return iframe;
}
exports.captureElement = captureElement;
function populateFrame(iframe, target, options) {
    lodash_1.toArray(target.children).forEach(function (node) {
        return iframe.contentDocument.body.appendChild(node);
    });
    addExternalStyles(iframe, target, options);
    FontLoader_1.loadFontsForFrame(iframe, options);
    target.parentElement.removeChild(target);
    iframe.setAttribute('style', null);
}
function addExternalStyles(iframe, target, options) {
    cloneParentStylesheets(iframe);
    loadAdditionalStylesheets(iframe, options);
    addInternalStyles(iframe);
}
function cloneParentStylesheets(iframe) {
    var links = lodash_1.toArray(document.getElementsByTagName('LINK')).filter(function (node) {
        return node.getAttribute('rel').toLowerCase() === 'stylesheet';
    });
    var stylesheets = lodash_1.toArray(document.getElementsByTagName('STYLE'));
    links.concat(stylesheets).map(function (node) {
        return iframe.contentDocument.head.appendChild(node.cloneNode(true));
    });
}
function addInternalStyles(iframe) {
    var stylesheet = void 0,
        css = void 0;
    stylesheet = iframe.contentDocument.createElement('style');
    iframe.contentDocument.head.appendChild(stylesheet);
    css = stylesheet.sheet;
    css.insertRule('html, body { display: flex !important; width: auto !important; height: auto !important; }', 0);
    css.insertRule('body { margin: 0 !important; overflow: hidden; }', 0);
}
function loadAdditionalStylesheets(iframe, options) {
    var stylesheets = lodash_1.get(options, 'stylesheets', []);
    lodash_1.forEach(stylesheets, function (url) {
        return loadAdditionalStylesheet(iframe, url);
    });
}
function loadAdditionalStylesheet(iframe, url) {
    var link = iframe.contentDocument.createElement('link');
    link.setAttribute('rel', 'stylesheet');
    link.setAttribute('type', 'text/css');
    link.setAttribute('media', 'all');
    link.setAttribute('href', url);
    iframe.contentDocument.head.appendChild(link);
}
exports.loadAdditionalStylesheet = loadAdditionalStylesheet;
