// kwill/lib/options/ValidateOptions.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var log = require("loglevel");
var lodash_1 = require("lodash");
var Find_1 = require("../utils/dom/element/Find");
function areOptionsValid(options) {
    if (!options) return;
    return isEditorSelectorValid(options);
}
exports.areOptionsValid = areOptionsValid;
function isEditorSelectorValid(options) {
    var element = void 0;
    try {
        element = Find_1.findOne(options.selector);
        return !!element;
    } catch (e) {
        log.warn('no editor element found for instance ID:', options.id);
        return false;
    }
}
exports.isEditorSelectorValid = isEditorSelectorValid;
function populateDefaultEditorOptions(options) {
    var defaults = void 0;
    defaults = {
        id: lodash_1.uniqueId(),
        style: {
            textAlign: 'left',
            lineHeight: '1',
            letterSpacing: '0px'
        },
        interaction: {
            show_popup_on_click: false,
            expand_x: true,
            expand_y: true
        }
    };
    return lodash_1.defaultsDeep(options, defaults);
}
exports.populateDefaultEditorOptions = populateDefaultEditorOptions;
function populateDefaultFormatUiOptions(options) {
    var defaults = void 0;
    defaults = {};
    return lodash_1.defaultsDeep(options, defaults);
}
exports.populateDefaultFormatUiOptions = populateDefaultFormatUiOptions;
