// kwill/src/ts/text-format-ui/createTextFormatUi.ts
import {areOptionsValid, populateDefaultEditorOptions, populateDefaultFormatUiOptions} from "../options/ValidateOptions";
import {oneTimeSetup} from "../globals/OneTimeSetup";
import {createTextFormatUi} from "./TextFormatUi";
import {TextFormatUiBase} from "./ui-components/TextFormatUiBase";

export interface IKwillFormatUiOptions
{
    popupSelector?: string | HTMLElement;
    toolbarSelector?: string | HTMLElement;
    fonts?:
        {
            google?: string[];
            glyphs?: string;
        },
    stylesheets?: string[];
}


/**
 * create a new instance of the Kwill editor,
 * using the targeted element as a base
 * @param options
 * @return {TextFormatUiBase}
 */
export function createKwillTextFormatUi(options: IKwillFormatUiOptions) : TextFormatUiBase
{
    //  do any one-time init we need to

    oneTimeSetup(options);

    //  add any missing default options

    populateDefaultFormatUiOptions(options);

    //  set up the text format UI for the new instance

    return createTextFormatUi(options);
}
