// kwil/dist/api_client/client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const base64_1 = require("../utils/base64");
const bytes_1 = require("../utils/bytes");
const api_1 = require("./api");
class Client extends api_1.Api {
    constructor(opts) {
        super(opts.kwilProvider, opts);
    }
    getFundingConfig() {
        const _super = Object.create(null, {
            get: { get: () => super.get }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield _super.get.call(this, `/api/v1/config`);
            return checkRes(res, r => r);
        });
    }
    getSchema(dbid) {
        const _super = Object.create(null, {
            get: { get: () => super.get }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield _super.get.call(this, `/api/v1/databases/${dbid}/schema`);
            return checkRes(res, r => r.dataset);
        });
    }
    getAccount(owner) {
        const _super = Object.create(null, {
            get: { get: () => super.get }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield _super.get.call(this, `/api/v1/accounts/${owner}`);
            return checkRes(res, r => r.account);
        });
    }
    listDatabases(owner) {
        const _super = Object.create(null, {
            get: { get: () => super.get }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield _super.get.call(this, `/api/v1/${owner}/databases`);
            return checkRes(res, r => r.databases);
        });
    }
    estimateCost(tx) {
        const _super = Object.create(null, {
            post: { get: () => super.post }
        });
        return __awaiter(this, void 0, void 0, function* () {
            let req = { tx };
            const res = yield _super.post.call(this, `/api/v1/estimate_price`, req);
            return checkRes(res, r => r.price);
        });
    }
    broadcast(tx) {
        const _super = Object.create(null, {
            post: { get: () => super.post }
        });
        return __awaiter(this, void 0, void 0, function* () {
            if (!tx.isSigned()) {
                throw new Error('Tx must be signed before broadcasting.');
            }
            let req = { tx };
            const res = yield _super.post.call(this, `/api/v1/broadcast`, req);
            checkRes(res);
            let body = null;
            if (res.data.receipt.body) {
                const uint8 = new Uint8Array((0, base64_1.base64ToBytes)(res.data.receipt.body));
                const decoder = new TextDecoder('utf-8');
                const jsonString = decoder.decode(uint8);
                body = JSON.parse(jsonString);
            }
            const cleanReceipt = !body ? {
                txHash: (0, bytes_1.Uint8ArrayToHex)((0, base64_1.base64ToBytes)(res.data.receipt.txHash)),
                fee: res.data.receipt.fee,
                body: null
            } : {
                txHash: (0, bytes_1.Uint8ArrayToHex)((0, base64_1.base64ToBytes)(res.data.receipt.txHash)),
                fee: res.data.receipt.fee,
                body: body
            };
            return {
                status: res.status,
                data: cleanReceipt
            };
        });
    }
    ping() {
        const _super = Object.create(null, {
            get: { get: () => super.get }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield _super.get.call(this, `/api/v1/ping`);
            return checkRes(res, r => r.message);
        });
    }
    selectQuery(query) {
        const _super = Object.create(null, {
            post: { get: () => super.post }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield _super.post.call(this, `/api/v1/query`, query);
            return checkRes(res, r => r.result);
        });
    }
}
exports.default = Client;
function checkRes(res, selector) {
    if (res.status != 200 || !res.data) {
        throw new Error(JSON.stringify(res.data) || 'An unknown error has occurred.  Please check your network connection.');
    }
    if (!selector) {
        return { status: res.status, data: undefined };
    }
    return {
        status: res.status,
        data: selector(res.data)
    };
}
