// kwill/lib/constants/Constants.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
exports.KWILL_CONSTS = {
    elements: {
        selectors: {
            text_pane: 'kwill-text-pane'
        }
    },
    events: {
        EDIT_START: 'edit start',
        EDIT_PROGRESS: 'edit progress',
        EDIT_COMPLETE: 'edit complete',
        TEXT_FORMAT: 'text format',
        CLEAR_SELECTION: 'clear selection',
        SHOW_TEXT_FORMAT_POPUP: 'show text format popup',
        HIDE_TEXT_FORMAT_POPUP: 'hide text format popup'
    },
    textFormats: {
        BOLD: 'bold',
        ITALIC: 'italic',
        UNDERLINE: 'underline',
        STRIKETHROUGH: 'strikethrough',
        TEXT_ALIGN: 'text-align',
        FONT_FAMILY: 'font-family',
        FONT_SIZE: 'font-size',
        LINE_HEIGHT: 'line-height',
        LETTER_SPACING: 'letter-spacing',
        TEXT_COLOUR: 'color',
        BACKGROUND_COLOUR: 'background-colour',
        CLASS: 'class',
        LINK: 'link',
        CLEAR_STYLE: 'clear-style'
    }
};
