// kwill/lib/editor/track-changes/TrackChanges.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var lodash_1 = require("lodash");
var Constants_1 = require("../../constants/Constants");
var GlobalEvents_1 = require("../../events/GlobalEvents");
var TextEditor_1 = require("../TextEditor");
var contentChangeDelay = void 0,
    contentChangeFlags = void 0;
function startTrackingEdits(instanceID) {
    var instanceState = void 0;
    instanceState = {
        id: instanceID,
        contentHasChanged: false,
        startHeight: getInstanceHeight(instanceID)
    };
    setInstanceState(instanceID, instanceState);
}
exports.startTrackingEdits = startTrackingEdits;
function trackEdits(instanceID) {
    getInstanceState(instanceID).contentHasChanged = true;
    if (contentChangeDelay) clearTimeout(contentChangeDelay);
    contentChangeDelay = window.setTimeout(function () {
        return GlobalEvents_1.getSignal(Constants_1.KWILL_CONSTS.events.EDIT_PROGRESS).dispatch(getInstanceState(instanceID));
    }, 1000);
}
exports.trackEdits = trackEdits;
function stopTrackingEdits() {
    if (contentChangeDelay) clearTimeout(contentChangeDelay);
    lodash_1.forIn(getContentStateDictionary(), function (instanceState, instanceID) {
        try {
            if (instanceState.contentHasChanged === true) {
                instanceState.endHeight = getInstanceHeight(instanceID);
                GlobalEvents_1.getSignal(Constants_1.KWILL_CONSTS.events.EDIT_COMPLETE).dispatch(instanceState);
                instanceState.contentHasChanged = false;
            }
        } catch (e) {}
    });
}
exports.stopTrackingEdits = stopTrackingEdits;
function getInstanceHeight(instanceID) {
    return TextEditor_1.getTextEditor(instanceID).height;
}
function getInstanceState(instanceID) {
    return getContentStateDictionary()[instanceID];
}
function setInstanceState(instanceID, instanceState) {
    getContentStateDictionary()[instanceID] = instanceState;
}
function getContentStateDictionary() {
    return contentChangeFlags || (contentChangeFlags = {});
}
