// kwill/lib/index.js
"use strict";

function __export(m) {
    for (var p in m) {
        if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
var Constants_1 = require("./constants/Constants");
exports.KWILL_CONSTS = Constants_1.KWILL_CONSTS;
var FontList_1 = require("./editor/info/format-sniffer/FontList");
exports.getAllFonts = FontList_1.getAllFonts;
var GlobalEvents_1 = require("./events/GlobalEvents");
exports.getSignal = GlobalEvents_1.getSignal;
var EditorInstanceUtils_1 = require("./editor/EditorInstanceUtils");
exports.getTextEditorContent = EditorInstanceUtils_1.getTextEditorContent;
exports.setTextEditorContent = EditorInstanceUtils_1.setTextEditorContent;
var TextFormatUiBase_1 = require("./text-format-ui/ui-components/TextFormatUiBase");
exports.TextFormatUiBase = TextFormatUiBase_1.TextFormatUiBase;
var TextEditor_1 = require("./editor/TextEditor");
exports.destroyTextEditor = TextEditor_1.destroyTextEditor;
var destroyAllInstances_1 = require("./editor/destroyAllInstances");
exports.destroyAllKwillInstances = destroyAllInstances_1.destroyAllKwillInstances;
__export(require("./editor/createKwillEditor"));
__export(require("./text-format-ui/createTextFormatUi"));
