// kwill/lib/KwillStatic.js
"use strict";

Object.defineProperty(exports, "__esModule", { value: true });
var ValidateOptions_1 = require("./options/ValidateOptions");
var OneTimeSetup_1 = require("./globals/OneTimeSetup");
var TextEditor_1 = require("./editor/TextEditor");
var GlobalEvents_1 = require("./events/GlobalEvents");
var FontList_1 = require("./editor/info/format-sniffer/FontList");
var FormatSniffer_1 = require("./editor/info/format-sniffer/FormatSniffer");
var GlobalInteractions_1 = require("./globals/GlobalInteractions");
var TextFormatUi_1 = require("./text-format-ui/TextFormatUi");
var EditorInstanceUtils_1 = require("./editor/EditorInstanceUtils");
function createKwillEditor(options) {
    if (!ValidateOptions_1.areOptionsValid(options)) return;
    OneTimeSetup_1.oneTimeSetup(options);
    ValidateOptions_1.populateDefaultEditorOptions(options);
    return TextEditor_1.createTextEditor(options.id, options);
}
exports.createKwillEditor = createKwillEditor;
function createKwillTextFormatUi(options) {
    OneTimeSetup_1.oneTimeSetup(options);
    ValidateOptions_1.populateDefaultFormatUiOptions(options);
    return TextFormatUi_1.createTextFormatUi(options);
}
exports.createKwillTextFormatUi = createKwillTextFormatUi;
exports.getTextFormat = FormatSniffer_1.sniffFormats;
exports.getKwillFontList = FontList_1.getAllFonts;
exports.getKwillSignal = function (id) {
    return GlobalEvents_1.getGlobalSignal(id);
};
exports.getKwillInstanceContent = function (id) {
    return EditorInstanceUtils_1.getTextEditorContent(id);
};
exports.setKwillInstanceContent = function (id, content) {
    return EditorInstanceUtils_1.setTextEditorContent(id, content);
};
function destroyAllKwillInstances(forever) {
    GlobalInteractions_1.removeGlobalListeners();
    GlobalEvents_1.destroyAllGlobalSignals();
    TextEditor_1.destroyAllTextEditors();
    TextFormatUi_1.destroyAllTextFormatUi();
    OneTimeSetup_1.destroyOneTimeSetup();
}
exports.destroyAllKwillInstances = destroyAllKwillInstances;
function destroyInstance(editorID) {
    TextEditor_1.destroyTextEditor(editorID);
}
exports.destroyInstance = destroyInstance;
var Constants_1 = require("./constants/Constants");
exports.kwillConstants = Constants_1.kwillConstants;
